BB-355 use updateDescription field to get object metadata in kafka lo…

welansari opened the pull request:
…g consumer

Issue: [BB-355](https://scality.atlassian.net/browse/BB-355)

**Problem:**
Some times oplog update events we get when using the Kafka log consumer don't have any metadata. This mostly happens when deleting the object right after the update.

The issue happens because we use the `fullDocument` field to retrieve the metadata, when reading the mongodb docs we see that the field represents a copy of the entire updated document at a point in time after the update, which means that in some cases this value can be unavailable like when deleting right after the update.

**Fix:**
Use the `updateDescription` field to get object metadata instead of fullDocument.
Values from `updateDescription` represent the exact changes that occurred to the object.
It is possible to have non complete metadata in this field but this shouldn't be an issue as we update the full object each time in the mongoClient.
One key updates like we do in the update we perform before each deletion will be ignored as we don't even have the `value` field that's supposed to contain the metadata.


[BB-355]: https://scality.atlassian.net/browse/BB-355?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[BB-355]: https://scality.atlassian.net/browse/BB-355?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Commit:
BB-355 use updateDescription field to get object metadata in kafka log consumer

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/backbeat/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-355 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.6.2`


Please check the `Fix Version/s` of BB-355, or the target
branch of this pull request.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-355 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.26`

* `8.5.5`

* `8.6.2`


Please check the `Fix Version/s` of BB-355, or the target
branch of this pull request.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/BB-355-support-delete-events-with-no-fulldocument-field` into
`development/8.4`
* [w/8.5/bugfix/BB-355-support-delete-events-with-no-fulldocument-field](https://github.com/scality/backbeat/compare/development/8.5...w/8.5/bugfix/BB-355-support-delete-events-with-no-fulldocument-field) will be merged into `development/8.5` 
* [w/8.6/bugfix/BB-355-support-delete-events-with-no-fulldocument-field](https://github.com/scality/backbeat/compare/development/8.6...w/8.6/bugfix/BB-355-support-delete-events-with-no-fulldocument-field) will be merged into `development/8.6` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/8.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

Commit:
BB-355 remove unused fullDocument field from oplog upate event

We no longer depend on the fullDocument field to get the updated
metadata of the objects. Removing it to reduce oplog event size

bert-e commented:
# History mismatch



Merge commit #39fcb3c0cf583fefa9a7e51b57d9c15837c237e2 on the integration branch
`w/8.5/bugfix/BB-355-support-delete-events-with-no-fulldocument-field` is merging a branch which is neither the current
branch `bugfix/BB-355-support-delete-events-with-no-fulldocument-field` nor the development branch
`development/8.5`.

It is likely due to a rebase of the branch `bugfix/BB-355-support-delete-events-with-no-fulldocument-field` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Kerkesni commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/BB-355-support-delete-events-with-no-fulldocument-field` into
`development/8.4`
* [w/8.5/bugfix/BB-355-support-delete-events-with-no-fulldocument-field](https://github.com/scality/backbeat/compare/development/8.5...w/8.5/bugfix/BB-355-support-delete-events-with-no-fulldocument-field) will be merged into `development/8.5` 
* [w/8.6/bugfix/BB-355-support-delete-events-with-no-fulldocument-field](https://github.com/scality/backbeat/compare/development/8.6...w/8.6/bugfix/BB-355-support-delete-events-with-no-fulldocument-field) will be merged into `development/8.6` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/8.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```









[BB-355]: https://scality.atlassian.net/browse/BB-355?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[BB-355]: https://scality.atlassian.net/browse/BB-355?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[BB-355]: https://scality.atlassian.net/browse/BB-355?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[BB-355]: https://scality.atlassian.net/browse/BB-355?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[BB-355]: https://scality.atlassian.net/browse/BB-355?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand commented on lib/queuePopulator/KafkaLogConsumer/ListRecordStream.js:
```diff
@@ -50,6 +50,29 @@ class ListRecordStream extends stream.Transform {
+        // document at a point in time after the update, this can cause issues
+        // when the object gets deleted just after the update (only for updates).
+        const updateDescriptionMd = changeStreamDocument.updateDescription
+            && changeStreamDocument.updateDescription.updatedFields &&
```
inconsistent wrapping: either put the operator at the end of the line (before line break) or at the beginning of next line (after line break), but do not mix...
(not sure what rule to apply, but at least it should be consistent)

francoisferrand commented on lib/queuePopulator/KafkaLogConsumer/ListRecordStream.js:
```diff
@@ -83,7 +106,7 @@ class ListRecordStream extends stream.Transform {
             entries: [{
                 key: changeStreamDocument.documentKey && changeStreamDocument.documentKey._id,
                 type: this._getType(changeStreamDocument.operationType, changeStreamDocument.fullDocument),
```
we still use fullDocument here : should be changed, e.g. pass `this._getObjectMd(changeStreamDocument)` instead to `getType()` ?

francoisferrand commented on lib/queuePopulator/KafkaLogConsumer/ListRecordStream.js:
```diff
@@ -50,6 +50,29 @@ class ListRecordStream extends stream.Transform {
+        if (changeStreamDocument.operationType === 'update') {
+            return updateDescriptionMd && JSON.stringify(updateDescriptionMd);
+        }
+        const fullDocumentMd = changeStreamDocument.fullDocument && changeStreamDocument.fullDocument.value;
```
`change.stream.full.document` is disabled, is it still relevant (e.g. in case 'set' operationType i guess) ?

welansari replied:
yes `insert` operations which we do when first creating an object only have the fullDocument field

francoisferrand commented on lib/queuePopulator/KafkaLogConsumer/ListRecordStream.js:
```diff
@@ -50,6 +50,29 @@ class ListRecordStream extends stream.Transform {
+            && changeStreamDocument.updateDescription.updatedFields &&
+            changeStreamDocument.updateDescription.updatedFields.value;
+        if (changeStreamDocument.operationType === 'update') {
+            return updateDescriptionMd && JSON.stringify(updateDescriptionMd);
```
are there not 2 cases, e.g. when we set the whole value vs when we update a single 'deep' field ?

From yesterday, i understand we update single field only when setting the 'delete' flag, so we may actually just skip that oplog entry?

welansari replied:
it will be skipped at the logReader level

welansari replied:
https://github.com/scality/backbeat/blob/d9aa80bfcc4335f65379223e4226c504da16ce52/lib/queuePopulator/LogReader.js#L496-L498

francoisferrand replied:
may be useful to add a comment in `getObjectMD()` explaining what we expect here (and how it will be fixed otherwise)

bert-e commented:
# History mismatch



Merge commit #c3767ebc161d6cd1a881604e349bff8670ad7736 on the integration branch
`w/8.5/bugfix/BB-355-support-delete-events-with-no-fulldocument-field` is merging a branch which is neither the current
branch `bugfix/BB-355-support-delete-events-with-no-fulldocument-field` nor the development branch
`development/8.5`.

It is likely due to a rebase of the branch `bugfix/BB-355-support-delete-events-with-no-fulldocument-field` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

welansari commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/BB-355-support-delete-events-with-no-fulldocument-field` into
`development/8.4`
* [w/8.5/bugfix/BB-355-support-delete-events-with-no-fulldocument-field](https://github.com/scality/backbeat/compare/development/8.5...w/8.5/bugfix/BB-355-support-delete-events-with-no-fulldocument-field) will be merged into `development/8.5` 
* [w/8.6/bugfix/BB-355-support-delete-events-with-no-fulldocument-field](https://github.com/scality/backbeat/compare/development/8.6...w/8.6/bugfix/BB-355-support-delete-events-with-no-fulldocument-field) will be merged into `development/8.6` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/8.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```









[BB-355]: https://scality.atlassian.net/browse/BB-355?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[BB-355]: https://scality.atlassian.net/browse/BB-355?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[BB-355]: https://scality.atlassian.net/browse/BB-355?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[BB-355]: https://scality.atlassian.net/browse/BB-355?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[BB-355]: https://scality.atlassian.net/browse/BB-355?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand reviewed (approved):
LGTM

XinLi0207 commented on lib/queuePopulator/KafkaLogConsumer/ListRecordStream.js:
```diff
@@ -50,6 +50,29 @@ class ListRecordStream extends stream.Transform {
+            return updateDescriptionMd && JSON.stringify(updateDescriptionMd);
+        }
+        const fullDocumentMd = changeStreamDocument.fullDocument && changeStreamDocument.fullDocument.value;
+        return fullDocumentMd && JSON.stringify(fullDocumentMd);
```
just a question out of my curiosity

the entries.value before it was `changeStreamDocument.fullDocument && JSON.stringify(changeStreamDocument.fullDocument.value),`

now we kind of changed it to sth like 
```
changeStreamDocument.fullDocument && 
changeStreamDocument.fullDocument.value &&
JSON.stringify(changeStreamDocument.fullDocument && 
changeStreamDocument.fullDocument.value)
```

what do we want to return here as value? why we don't just let the last one as return value

welansari replied:
Looking at it, there is a redundant check indeed, `JSON.stringify(undefined)` should equal `undefined`.
So this `return fullDocumentMd && JSON.stringify(fullDocumentMd);` should become `return JSON.stringify(fullDocumentMd);`

XinLi0207 replied:
ok cool! you already updated that 👍

bert-e commented:
# History mismatch



Merge commit #f307bf239a10e4942c9b0bd906d6c4dc0d869383 on the integration branch
`w/8.5/bugfix/BB-355-support-delete-events-with-no-fulldocument-field` is merging a branch which is neither the current
branch `bugfix/BB-355-support-delete-events-with-no-fulldocument-field` nor the development branch
`development/8.5`.

It is likely due to a rebase of the branch `bugfix/BB-355-support-delete-events-with-no-fulldocument-field` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Kerkesni commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/BB-355-support-delete-events-with-no-fulldocument-field` into
`development/8.4`
* [w/8.5/bugfix/BB-355-support-delete-events-with-no-fulldocument-field](https://github.com/scality/backbeat/compare/development/8.5...w/8.5/bugfix/BB-355-support-delete-events-with-no-fulldocument-field) will be merged into `development/8.5` 
* [w/8.6/bugfix/BB-355-support-delete-events-with-no-fulldocument-field](https://github.com/scality/backbeat/compare/development/8.6...w/8.6/bugfix/BB-355-support-delete-events-with-no-fulldocument-field) will be merged into `development/8.6` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/8.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

welansari commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.4`

* :heavy_check_mark: `development/8.5`

* :heavy_check_mark: `development/8.6`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/8.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.4`

* :heavy_check_mark: `development/8.5`

* :heavy_check_mark: `development/8.6`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/8.3`



Please check the status of the associated issue BB-355.

Goodbye kerkesni.
